# Week 3 Exercise: Predicting Numerical Target Variable using Multiple Linear Regression

This notebook will use a dataset from a previous assignment. This dataset has already been through some basic cleaning and transformation steps. The purpose of this notebook will be to use Multiple Linear Regression to create a model that will predict the target variable, 'y'.

## 1. Import Libraries

At this time, Pandas and NumPy libraries will be imported. scikit learn libraries will be imported as they are needed.

In [1]:
import pandas as pd
import numpy as np

## 2. Load the Data

In [2]:
df = pd.read_csv('eda_data_cleaned.csv')

## 3. Define the Predictors and Target

The target variable to predict is 'y'. The predictor variables are all the remaining variables in the dataset. In the next code cell I will separate X and y so the model knows which features belong to which group. 

In [4]:
X = df.drop('y', axis=1)
y = df['y']

## 4. Train-test Split

Data will be split into training data and testing data. This is to make sure the model is being tested with data that it has not seen before and that already has values for 'y' so that we can assess accuracy. 

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.1, random_state=123
)

## 5. Create the MLR Model

MLR will be used to predict the target variable, 'y', using the predictor variables in the the training dataset. This step will use the 'LinearRegression' estimator from scikit-learn to create the model. I am following the steps listed in the textbook.

1. Import the estimator class
2. Create an instance of the class
3. Use the fit() method of the instance to train the model
4. Use the predict() method to get predictions

I will also examine the model coefficients. 

In [11]:
from sklearn.linear_model import LinearRegression

ml_reg = LinearRegression()

ml_reg.fit(X_train, y_train)

y_pred_ml_reg = ml_reg.predict(X_train)

pd.Series(
    ml_reg.coef_,
    index=X_train.columns
).sort_values(ascending=False).round(6)

x10    21.719174
x13     0.000297
x5      0.000276
x4      0.000137
x3      0.000115
x9      0.000075
x11     0.000032
x6     -0.000000
x2     -0.000005
x0     -0.000005
x7     -0.000023
x8     -0.000030
x12    -0.000131
x1     -0.000165
dtype: float64

## 6. Evaluate Predictions Using Mean Squared Error

MSE measures the average squared difference between the actual target values and the model's predictions. The smaller the MSE, the closer the predictions are to the actual values. For this step, sklearn.metrics library 'mean_squared_error' will be imported.

In [15]:
from sklearn.metrics import mean_squared_error

mse_ml_reg = mean_squared_error(
    y_true=y_train,
    y_pred=y_pred_ml_reg
)

mse_ml_reg

0.0830605189808644

## 7. Calculate Null Model MSE

The null model will provide a baseline for evaluating the MLR model using MSE. The MSE of the null model and the MSE of the MLR model created will be compared.

In [14]:
y_pred_null_model = np.full(y_train.shape, y_train.mean())
mse_null_model = mean_squared_error(
    y_true=y_train,
    y_pred=y_pred_null_model)
mse_null_model

0.08313357011331682

## 8. Compare MLR and Null Model

The MLR model produced an MSE of approximately 0.08306, while the null model produced an MSE of approximately 0.08313. The MLR model MSE is slightly smaller which indicates the MLR model's predictions are closer to the actual values. Since the difference between the two values is very minimal, it suggests that the predictor variables only provide a small improvement to the prediction values over simply caluclating the mean of the target variable for each observation. 

## 9. Test Predictions

The trained MLR model will be used to make predictions using the test dataset. 

In [16]:
y_pred_ml_reg_test = ml_reg.predict(X_test)

## 10. Evaluate the Model Using Test Data

The MSE will be calculated for the predictions on the test dataset. This will provide an idea of accuracy for how the trained MLR model performs with data that was not used for training. 

In [17]:
mse_ml_reg_test = mean_squared_error(
    y_true=y_test,
    y_pred=y_pred_ml_reg_test
)

mse_ml_reg_test

0.0840542229910024

### Interpretation/Evaluation of Model on Test Data

The training MSE was approximately 0.08306, while the test MSE was approximately 0.08405. The test MSE is slightly higher. The very small difference suggests that the model performed similarly with both datasets. 